WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java

```diff
@@ -778,7 +778,8 @@ private int getCurrentPostStatusIndex() {
     // Post Format Helpers
 
     private void updatePostFormatKeysAndNames() {
-        if (getActivity() == null || getSite() == null) {
+        final SiteModel site = getSite();
+        if (!isAdded() || site == null) {
```

Comment:
I don't think we really care whether the fragment is currently added since we are not doing any view manipulation. Just a site `null` check seems enough to me since `getSite()` will check for the activity being `null`.

```suggestion
        if (site == null) {
```

Reply:
True, there are no View manipulations in the method. I'd be willing to keep the `isAdded()` check only as an small optimization since the method does update some Fragment-local state, which is unneeded to do if the Fragment is not added. Hardy strong feelings on this one though.

I kept the relevant change as a separate commit anyway and happy to revert it if you still feel like removing the check. Let me know 👍

Reply:
Actually, I think we want to make sure we handle this at all times, since it's from a callback and we need that information. It's hardly ideal, but we didn't have `ViewModel`s when we wrote all this, so we have to live with it for now.

Reply:
Good point, will leave it as is with the `isAdded()`.

Reply:
Umm.. I think my comment was not clear, I think we need to handle it even if the fragment is not added at the time of the callback. Sorry for the confusion.

Reply:
Oh, I see :), no prob. Lemme give it a look and remove the isAdded check. Will probably also add some comment about it.

Reply:
Hmm, there's a `getResources()` call in line 787 that I think will throw if the fragment is not added.

Reply:
Ha, I missed that in the review, good point. I guess that's why we had the `getActivity()` check before. I don't think it'd throw if it's not added though, it'd throw if the `host` is null, so I feel `getActivity()` check is still better.

A little bit of context into my thinking. I have been recently thinking about the `isAdded` checks  we have all around the app and I feel like they are just masking problems. Ideally a fragment would stop receiving callbacks once it's paused (or stopped) and the `isAdded` checks wouldn't be there for anything. We can't really fix the root cause of these `isAdded` checks for every fragment right now, but I am reluctant to use it when it's not absolutely necessary. Even a little bit of improvement is good in my opinion.

In this case, if all we need is the `resources` I think we should stick to the activity check. Having said all this and the previous comments, I don't feel too strongly about any of this for an old fragment. I am OK to go with whatever you decide at this point, just thought I'd provide all the context necessary for you to make a decision :)

Reply:
Thanks for elaborating in your thinking Oguz. And I totally agree with the feeling about `isAdded()`. It's been long time now that we have sprinkled the app with it and had started to feel like snake oil :(

So, here's the thing: this originating issue points to some timing issue (which still feels weird to happen as often as it does to be honest, but anyway) so, just reinstating the `getActivity()` check doesn't feel like a safest route.

Here's what I'm thinking to do at this point:
1. Extract the initialization of the `mPostFormatKeys` and `mPostFormatNames` to an earlier point in Fragment's lifecycle and use it as cache, plus
2. only leave the `site` null check in the method

How does that sound?

Reply:
> Extract the initialization of the `mPostFormatKeys` and `mPostFormatNames` to an earlier point in Fragment's lifecycle and use it as cache, plus

Can we do this? If I am not mistaken, it's called in two places, one in `onActivityCreated()` the other in `onPostFormatsChanged()`. The `onActivityCreated` is one of the first places we could do it and `onPostFormatsChanged` is a FluxC callback. Maybe I am missing something, or didn't fully understand your suggestion?

> So, here's the thing: this originating issue points to some timing issue (which still feels weird to happen as often as it does to be honest, but anyway) so, just reinstating the getActivity() check doesn't feel like a safest route.

I think the issue was just that we were using `getSite` and by caching it with `final` at the beginning of the method was the necessary fix for this issue. I might be wrong though :)

Reply:
Pushed d0dc7292f1 that has what I had in mind. The caching of the lists happens in `onCreate()`, quite early that is, and then using it to initialize mPostFormatKeys and mPostFormatNames.

It starts to feel like over-engineering at this point, but maybe not too much. WDYT?

Reply:
That works for me @hypest, just tested it as well and once that extra empty line in `Line 156` is removed, I think we can merge this in 👍

Reply:
Done with 35c5ba9.